File: ds-leetcode/dp_minimum-falling-path-sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<vector<int>> dp(m, vector<int>(n, 0));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0) {
                    dp[i][j] = matrix[i][j];
                } else {
                    int min_num = INT_MAX;
                    for (int k = 0; k < n; k++) {
                        if (k == j) continue;
                        min_num = min(min_num, dp[i - 1][k]);
                    }
                    dp[i][j] = min_num + matrix[i][j];
                }
            }
        }
        int res = INT_MAX;
        for (int i = 0; i < n; i++) {
            res = min(res, dp[m - 1][i]);
        }
        return res;
    }
    int min(int a, int b) {
        return a < b ? a : b;
    }
};

File: ds-leetcode/dp_minimum-falling-path-sum.py
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        if not matrix:
            return 0
        m, n = len(matrix), len(matrix[0])
        dp = [[0] * n for _ in range(m)]
        dp[0] = matrix[0]
        for i in range(1, m):
            for j in range(n):
                dp[i][j] = matrix[i][j] + min(dp[i - 1][(j - 1) % n], dp[i - 1][(j + 1) % n], dp[i - 1][j])
        return min(dp[m - 1])